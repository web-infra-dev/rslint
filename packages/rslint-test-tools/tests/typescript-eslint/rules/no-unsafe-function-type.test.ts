import { RuleTester, getFixturesRootDir } from '../RuleTester.ts';

const rootPath = getFixturesRootDir();

const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootPath,
    },
  },
});

ruleTester.run('no-unsafe-function-type', {
  valid: [
    'let value: () => void;',
    'let value: <T>(t: T) => T;',
    `
      // create a scope since it's illegal to declare a duplicate identifier
      // 'Function' in the global script scope.
      {
        type Function = () => void;
        let value: Function;
      }
    `,
  ],
  invalid: [
    {
      code: 'let value: Function;',
      errors: [
        {
          column: 12,
          line: 1,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: 'let value: Function[];',
      errors: [
        {
          column: 12,
          line: 1,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: 'let value: Function | number;',
      errors: [
        {
          column: 12,
          line: 1,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: `
        class Weird implements Function {
          // ...
        }
      `,
      errors: [
        {
          column: 32,
          line: 2,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: `
        interface Weird extends Function {
          // ...
        }
      `,
      errors: [
        {
          column: 33,
          line: 2,
          messageId: 'bannedFunctionType',
        },
      ],
    },
  ],
});