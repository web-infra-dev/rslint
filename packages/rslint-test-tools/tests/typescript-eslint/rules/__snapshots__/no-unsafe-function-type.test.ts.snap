// Rstest Snapshot v1

exports[`no-unsafe-function-type > invalid 1`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "The \`Function\` type accepts any function-like value.
Prefer explicitly defining any function parameters and return type.",
      "messageId": "bannedFunctionType",
      "range": {
        "end": {
          "column": 20,
          "line": 1,
        },
        "start": {
          "column": 12,
          "line": 1,
        },
      },
      "ruleName": "no-unsafe-function-type",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-unsafe-function-type > invalid 2`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "The \`Function\` type accepts any function-like value.
Prefer explicitly defining any function parameters and return type.",
      "messageId": "bannedFunctionType",
      "range": {
        "end": {
          "column": 20,
          "line": 1,
        },
        "start": {
          "column": 12,
          "line": 1,
        },
      },
      "ruleName": "no-unsafe-function-type",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-unsafe-function-type > invalid 3`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "The \`Function\` type accepts any function-like value.
Prefer explicitly defining any function parameters and return type.",
      "messageId": "bannedFunctionType",
      "range": {
        "end": {
          "column": 20,
          "line": 1,
        },
        "start": {
          "column": 12,
          "line": 1,
        },
      },
      "ruleName": "no-unsafe-function-type",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-unsafe-function-type > invalid 4`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "The \`Function\` type accepts any function-like value.
Prefer explicitly defining any function parameters and return type.",
      "messageId": "bannedFunctionType",
      "range": {
        "end": {
          "column": 40,
          "line": 2,
        },
        "start": {
          "column": 32,
          "line": 2,
        },
      },
      "ruleName": "no-unsafe-function-type",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-unsafe-function-type > invalid 5`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "The \`Function\` type accepts any function-like value.
Prefer explicitly defining any function parameters and return type.",
      "messageId": "bannedFunctionType",
      "range": {
        "end": {
          "column": 41,
          "line": 2,
        },
        "start": {
          "column": 33,
          "line": 2,
        },
      },
      "ruleName": "no-unsafe-function-type",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;
